Explicit TCP/UDP socket binding, local port retrieval.

Faless opened the pull request:
Work on salvaging #43174 .

In this PR:

- Allow explicit `bind` of sockets used by `PacketPeerUDP` and `StreamPeerTCP`.
- Add `get_local_port` to TCP/UDP classes to retrieve local port.
- Rename `PacketPeerUDP.listen` to `PacketPeerUDP.bind` to make it consistent with TCP, and separate the concept of client/server (abstracted by `UDPServer`) and socket binding.

Commit:
Allow local port control on net_socket connections

Faless committed:
This renames PacketPeerUDP.listen to bind.

Faless commented:
Should be good to go when CI passes.

Calinou commented on core/io/packet_peer_udp.cpp:
```diff
@@ -159,9 +159,10 @@ int PacketPeerUDP::get_max_packet_size() const {
+Error PacketPeerUDP::bind(int p_port, const IP_Address &p_bind_address, int p_recv_buffer_size) {
 	ERR_FAIL_COND_V(!_sock.is_valid(), ERR_UNAVAILABLE);
 	ERR_FAIL_COND_V(_sock->is_open(), ERR_ALREADY_IN_USE);
+	ERR_FAIL_COND_V(p_port < 0 || p_port > 65535, ERR_INVALID_PARAMETER);
```
To match the other `ERR_FAIL_COND_V_MSG` that was added below:

```suggestion
	ERR_FAIL_COND_V_MSG(p_port < 1 || p_port > 65535, ERR_INVALID_PARAMETER, "The server port number must be between 1 and 65535 (inclusive).");
```

Calinou commented on core/io/packet_peer_udp.cpp:
```diff
@@ -210,6 +211,7 @@ Error PacketPeerUDP::connect_to_host(const IP_Address &p_host, int p_port) {
 	ERR_FAIL_COND_V(udp_server, ERR_LOCKED);
 	ERR_FAIL_COND_V(!_sock.is_valid(), ERR_UNAVAILABLE);
 	ERR_FAIL_COND_V(!p_host.is_valid(), ERR_INVALID_PARAMETER);
+	ERR_FAIL_COND_V_MSG(p_port < 1 || p_port > 65535, ERR_INVALID_PARAMETER, "The server port number must be between 0 and 65535 (inclusive).");
```
```suggestion
	ERR_FAIL_COND_V_MSG(p_port < 1 || p_port > 65535, ERR_INVALID_PARAMETER, "The server port number must be between 1 and 65535 (inclusive).");
```

Faless replied:
`0` is correct here,

Faless replied:
No, `1` is correct here, got tricked by GH partial diff. I'll fix the message.

Calinou commented on core/io/stream_peer_tcp.cpp:
```diff
@@ -67,21 +67,40 @@ void StreamPeerTCP::accept_socket(Ref<NetSocket> p_sock, IP_Address p_host, uint
 	ERR_FAIL_COND_V(!_sock.is_valid(), ERR_UNAVAILABLE);
 	ERR_FAIL_COND_V(_sock->is_open(), ERR_ALREADY_IN_USE);
-	ERR_FAIL_COND_V(!p_host.is_valid(), ERR_INVALID_PARAMETER);
+	ERR_FAIL_COND_V_MSG(p_port < 0 || p_port > 65535, ERR_INVALID_PARAMETER, "The local port number must be between 0 and 65535 (inclusive).");
```
```suggestion
	ERR_FAIL_COND_V_MSG(p_port < 1 || p_port > 65535, ERR_INVALID_PARAMETER, "The local port number must be between 1 and 65535 (inclusive).");
```

Calinou commented on core/io/tcp_server.cpp:
```diff
@@ -34,13 +34,15 @@ void TCP_Server::_bind_methods() {
 Error TCP_Server::listen(uint16_t p_port, const IP_Address &p_bind_address) {
 	ERR_FAIL_COND_V(!_sock.is_valid(), ERR_UNAVAILABLE);
 	ERR_FAIL_COND_V(_sock->is_open(), ERR_ALREADY_IN_USE);
+	ERR_FAIL_COND_V(p_port < 0 || p_port > 65535, ERR_INVALID_PARAMETER);
```
```suggestion
	ERR_FAIL_COND_V_MSG(p_port < 1 || p_port > 65535, ERR_INVALID_PARAMETER, "The server port number must be between 1 and 65535 (inclusive).");
```

Faless replied:
`0` is a valid port for listen/bind. I'll add the `_MSG` though.

Calinou commented on doc/classes/StreamPeerTCP.xml:
```diff
@@ -9,6 +9,18 @@
+			<argument index="1" name="host" type="String" default="&quot;*&quot;">
+			</argument>
+			<description>
+				Open the TCP socket, and bind it to the specifed local address.
```
```suggestion
				Opens the TCP socket and binds it to the specified local address.
```

Calinou commented on doc/classes/StreamPeerTCP.xml:
```diff
@@ -9,6 +9,18 @@
+			</argument>
+			<description>
+				Open the TCP socket, and bind it to the specifed local address.
+				This method is generally not needed, and only used to force the subsequent call to [method connect_to_host] to use the specified [code]host[/code] and [code]port[/code] as source address, which can be desired in some NAT punch through techniques, or when forcing the source network interface.
```
```suggestion
				This method is generally not needed, and only used to force the subsequent call to [method connect_to_host] to use the specified [code]host[/code] and [code]port[/code] as source address. This can be desired in some NAT punchthrough techniques, or when forcing the source network interface.
```

Calinou commented on doc/classes/PacketPeerUDP.xml:
```diff
@@ -9,11 +9,27 @@
 			</return>
 			<description>
-				Closes the UDP socket the [PacketPeerUDP] is currently listening on.
+				Closes underlying UDP socket the [PacketPeerUDP].
```
```suggestion
				Closes the [PacketPeerUDP]'s underlying UDP socket.
```

Calinou commented on core/io/udp_server.cpp:
```diff
@@ -89,6 +90,7 @@ Error UDPServer::poll() {
 Error UDPServer::listen(uint16_t p_port, const IP_Address &p_bind_address) {
 	ERR_FAIL_COND_V(!_sock.is_valid(), ERR_UNAVAILABLE);
 	ERR_FAIL_COND_V(_sock->is_open(), ERR_ALREADY_IN_USE);
+	ERR_FAIL_COND_V(p_port < 0 || p_port > 65535, ERR_INVALID_PARAMETER);
```
```suggestion
	ERR_FAIL_COND_V_MSG(p_port < 1 || p_port > 65535, ERR_INVALID_PARAMETER, "The server port number must be between 1 and 65535 (inclusive).");
```

Faless replied:
`0` is a valid port for `listen/bind`. I'll add the `_MSG` though.

Faless commented:
Addressed the issues.
Changed the error message to consistently reference either `local` or `remote` port (trying to make explicit the reason when `0` is allowed).

akien-mga commented:
Thanks!
